lasagne/utils.py

```diff
@@ -204,3 +204,76 @@ def compute_norms(array, norm_axes=None):
     norms = np.sqrt(np.sum(array**2, axis=sum_over))
 
     return norms
+
+
+def create_param(spec, shape, name=None):
+    """
+    Helper method to create Theano shared variables for layer parameters
+    and to initialize them.
+
+    Parameters
+    ----------
+
+    spec : numpy array, Theano shared variable, or callable
+        One of three things:
+            * a numpy array with the initial parameter values
+            * a Theano shared variable representing the parameters
+            * a function or callable that takes the desired shape of
+              the parameter array as its single argument.
+
+    shape : tuple
+        a tuple of integers representing the desired shape of the
+        parameter array.
+
+    name : string, optional
+        If a new variable is created, the name to give to the parameter
+        variable. This is ignored if `spec` is already a Theano shared
+        variable.
+
+    Returns
+    -------
+    Theano shared variable
+        a Theano shared variable representing layer parameters. If a
+        numpy array was provided, the variable is initialized to
+        contain this array. If a shared variable was provided, it is
+        simply returned. If a callable was provided, it is called, and
+        its output is used to initialize the variable.
+
+    Notes
+    -----
+    This method should be used in the constructor when creating a
+    :class:`Layer` subclass that has parameters. This enables the layer to
+    support initialization with numpy arrays, existing Theano shared
+    variables, and callables for generating initial parameter values.
+    """
+    if isinstance(spec, theano.compile.SharedVariable):
+        # We cannot check the shape here, the shared variable might not be
+        # initialized correctly yet. We can check the dimensionality
+        # though. Note that we cannot assign a name here. We could assign
+        # to the `name` attribute of the shared variable, but we shouldn't
+        # because the user may have already named the variable and we don't
+        # want to override this.
+        if spec.ndim != len(shape):
+            raise RuntimeError("shared variable has %d dimensions, "
+                               "should be %d" % (spec.ndim, len(shape)))
+        return spec
+
+    elif isinstance(spec, np.ndarray):
+        if spec.shape != shape:
+            raise RuntimeError("parameter array has shape %s, should be "
+                               "%s" % (spec.shape, shape))
+        return theano.shared(spec, name=name)
+
+    elif hasattr(spec, '__call__'):
+        arr = spec(shape)
+        if not isinstance(arr, np.ndarray):
+            raise RuntimeError("cannot initialize parameters: the "
+                               "provided callable did not return a numpy "
+                               "array")
```

Comment:
This is independent of this PR, but as I saw it now, this check should probably go something like this instead:
```python
arr = spec(shape)
try:
    arr = floatX(arr)
except Exception:
    ... did not return an array-like ...
if arr.shape != shape:
    ... did not return correct shape ...
return theano.shared(arr, name=name)
```
This should also handle scalars correctly.